firmware/usbarmory/src/dcp/aes128.rs

```diff
@@ -0,0 +1,215 @@
+use core::{
+    marker::PhantomData,
+    ptr::NonNull,
+    sync::atomic::{self, AtomicBool, Ordering},
+};
+
+use arrayref::array_ref;
+use block_cipher_trait::{
+    generic_array::{typenum::consts, GenericArray},
+    BlockCipher,
+};
+use pac::hw_dcp::HW_DCP;
+
+use crate::{
+    dcp::{
+        self, CipherMode, CipherSelect, Cmd, Control0, Control1, KeySelect, AES128_CHANNEL,
+        STAT_ERROR_MASK,
+    },
+    memlog, memlog_flush_and_reset, util,
+};
+
+/// AES-128 channel
+pub struct Aes128 {
+    key: KeySelect,
+    // this struct implicitly owns a shared reference to `HW_DCP` ..
+    _not_send_or_sync: PhantomData<*mut ()>,
+}
+
+// .. however it accesses different registers so instances are OK to `Send`
+unsafe impl Send for Aes128 {}
+
+impl Drop for Aes128 {
+    fn drop(&mut self) {
+        AES_IN_USE.store(false, Ordering::Release)
+    }
+}
+
+impl BlockCipher for Aes128 {
+    type KeySize = consts::U16;
+    type BlockSize = consts::U16;
+    type ParBlocks = consts::U1;
+
+    fn new(key: &GenericArray<u8, consts::U16>) -> Self {
+        Self::new_ram(key).expect("the `Aes128` is currently in use")
+    }
+
+    fn decrypt_block(&self, block: &mut GenericArray<u8, consts::U16>) {
+        self.xcrypt(block, false)
+    }
+
+    fn encrypt_block(&self, block: &mut GenericArray<u8, consts::U16>) {
+        self.xcrypt(block, true)
+    }
+}
+
+static AES_IN_USE: AtomicBool = AtomicBool::new(false);
+
+impl Aes128 {
+    /// Gets a handle to the AES-128 channel and configures it to use the unreadable OTP key
+    ///
+    /// This function returns `None` if the channel is currently in use
+    ///
+    /// **WARNING!** This routine does NOT check if the OTP was set to an all-zeros value
+    pub fn new_otp() -> Option<Self> {
+        Self::new_hardware(true)
+    }
+
+    /// Gets a handle to the AES-128 channel and configures it to use the unreadable UNIQUE key
+    ///
+    /// This function returns `None` if the channel is currently in use
+    ///
+    /// This UNIQUE key is generated from the OTP key and a device-specific 64-bit value
+    pub fn new_unique() -> Option<Self> {
+        Self::new_hardware(false)
+    }
+
+    // Creates a cipher that uses one of the two available hardware keys
+    fn new_hardware(otp: bool) -> Option<Self> {
```

Comment:
Tiny nit on an argument to an internal function, but I think this would make more sense as...

```suggestion
    fn new_hardware(unique: bool) -> Option<Self> {
```

Reply:
changed the argument to an enum; they read better than booleans